tests/main/snap-quota-services/task.yaml

```diff
@@ -0,0 +1,76 @@
+    # old systemd versions, we need at least 230 to avoid buggy slice usage
+    # reporting
+
+    snap set-quota foobar --memory=1MB 2>&1 | tr '\n' ' ' | tr -s ' ' | MATCH "systemd version 2[0-2][0-9] is too old \(expected at least 230\)"
+    exit 0
+  fi
+
+  # Create the top level snap group with a memory limit. test-snapd-stressd
+  # contains 4 services, which makes it perfect for our testing as we only need
+  # this one snap to test isolating services.
+  snap set-quota test-top --memory=400MB test-snapd-stressd
+
+  echo "Create a sub-level group for the stress service and put it into that"
+  snap set-quota test-sub --parent=test-top --cpu=10% test-snapd-stressd.stress-sc
+
+  echo "Test the quota groups listings look sane"
+  snap quotas | cat -n > quotas.txt
+  MATCH "     1\s+Quota\s+Parent\s+Constraints\s+Current$" < quotas.txt
+  MATCH "     2\s+test-top\s+memory=400MB(\s*|\s*memory=[0-9.a-zA-Z]+)+$" < quotas.txt
+  MATCH "     3\s+test-sub\s+test-top\s+cpu=10%\s+$" < quotas.txt
+
+  echo "Test that service details are listed correctly in test-sub"
+  snap quota test-sub | cat -n > details.txt
+  MATCH "     1\s+name:\s+test-sub$" < details.txt
+  MATCH "     2\s+parent:\s+test-top$" < details.txt
+  MATCH "     3\s+constraints:$" < details.txt
+  MATCH "     4\s+cpu-count:\s+0$" < details.txt
+  MATCH "     5\s+cpu-percentage:\s+10$" < details.txt
+  MATCH "     6\s+current:$" < details.txt
+  MATCH "     7\s+services:$" < details.txt
+  MATCH "     8\s+-\s+test-snapd-stressd.stress-sc$" < details.txt
+
+  echo "Test that it fails when we apply a journal quota to the sub-group"
+  snap set-quota test-sub --journal-size=128MB 2>&1 | tr '\n' ' ' | tr -s ' ' | MATCH 'error: cannot update quota group: cannot update group "test-sub": journal quotas are not supported for individual services'
+
+  echo "Test that we can however put a journal quota on the parent group"
+  snap set-quota test-top --journal-size=128MB
+
+  ls /nono
+
+  # Next up is checking for slices and unit contents
+  # For test-snapd-stressd.logger (which is another service in test-snapd-stressd)
+  # we will expect the unit to mention the top group slice, but for test-snapd-stressd.stress-sc
+  # we will expect the unit to mention the sub group slice.
+  TOP_SLICENAME="snap.$(systemd-escape --path test-top).slice"
+  SUB_SLICENAME="snap.$(systemd-escape --path test-top/test-sub).slice"
+
+  echo "Verify that one of the non stress-sc services mention the top group"
+  systemctl show --property=Slice snap.test-snapd-stressd.logger.service | grep -F "Slice=$TOP_SLICENAME"
+
+  echo "Verify that stress-sc service mention the sub group"
+  systemctl show --property=Slice snap.test-snapd-stressd.stress-sc.service | grep -F "Slice=$SUB_SLICENAME"
+
+  # echo "Verify that stress-sc service mention the journal namespace"
+  # systemctl show --property=LogNamespace snap.test-snapd-stressd.stress-sc.service | MATCH "LogNamespace=snap-test-sub"
+
+  snap remove-quota test-sub
```

Comment:
(nitpick) I wonder if we should use `tests.cleanup defer` closer to where the `test-sub` and `test-top` is used but it's a minor quibble.

Reply:
This is a really good suggestion

Reply:
I need to do some followups on this anyway from the previous PRs so I'll add this

Reply:
Fixed this anyway!